libs/analysis/bbhnet/analysis/vt.py

```diff
@@ -0,0 +1,166 @@
+        recovered_parameters:
+            Dictionary of recovered parameters
+        n_injections:
+            Number of total injections
+        livetime:
+            Livetime in seconds over which injections were performed
+        cosmology:
+            Astropy Cosmology object used for volume calculation
+    """
+
+    source: "bilby.core.prior.PriorDict"
+    recovered_parameters: Dict[str, np.ndarray]
+    n_injections: int
+    livetime: float
+    cosmology: "Cosmology" = cosmo.Planck15
+
+    def __post_init__(self):
+        # convert recovered parameters to a list of dictionaries
+        self.recovered_parameters = [
+            dict(zip(self.recovered_parameters, col))
+            for col in zip(*self.recovered_parameters.values())
+        ]
+        dl_prior = self.source["luminosity_distance"]
+        dl_min, dl_max = [dl_prior.minimum, dl_prior.maximum]
+
+        # if the source distribution has a dec prior,
+        # use it to calculate the area on the sky
+        # over which injections have been made
+        # otherwise, calculate_astrophysical_volume assumes the full sky
+        if "dec" in self.source:
+            dec_prior = self.source["dec"]
+            dec_min, dec_max = dec_prior.minimum, dec_prior.maximum
+        else:
+            dec_min = dec_max = None
+
+        # calculate the astrophysical volume over
+        # which injections have been made.
+        self.volume = calculate_astrophysical_volume(
+            dl_min=dl_min,
+            dl_max=dl_max,
+            dec_min=dec_min,
+            dec_max=dec_max,
+            cosmology=self.cosmology,
+        )
+
+    def weights(self, target: Optional["bilby.core.prior.PriorDict"] = None):
+        """
+        Calculate the weights for the samples.
+        """
+
+        # if no target distribution is passed,
+        # use the source distribution
+        if target is None:
+            target = self.source
+
+        weights = []
+        for sample in self.recovered_parameters:
+            # calculate the weight for each sample
+            # using the source and target distributions
+            weight = target.prob(sample) / self.source.prob(sample)
+            weights.append(weight)
+
+        return np.array(weights)
+
+    def calculate_vt(
+        self,
+        target: Optional["bilby.core.prior.PriorDict"] = None,
+    ):
+        """
+        Calculates the VT and its uncertainty. See equations
+        8 and 9 in https://arxiv.org/pdf/1904.10879.pdf
+
+        Args:
+            target:
+                Bilby PriorDict of the target distribution
+                used for importance sampling. If None, the source
+                distribution is used.
+
+        Returns tuple of (vt, std, n_eff)
+        """
+        weights = self.weights(target)
+
+        mu = np.sum(weights) / self.n_injections
+
+        v0 = self.livetime * YEARS_PER_SECOND * self.volume
+        vt = mu * v0
+        variance = v0**2 * (
```

Comment:
I might suggest breaking this up into a couple lines to help make it more clear but not a big deal